### ISSUE TITLE
Cannot read property 'trim' of undefined.

### ISSUE THREAD

**Issue by matthiaswatkins** [npm/npm#7984]
I'm a complete newbie, trying to run an install from a downloaded [package.json](https://github.com/sequitur/raconteur/blob/master/package.json). I get this error on a stdout.trim() call inside add-remote-git.js. This install worked yesterday on another PC (both PCs are Windows 7 machines), but is giving me fits today. Here are some possibly relevant lines from the [log file](https://gist.github.com/matthiaswatkins/41d539f72b7ce3d46231):

```
303 verbose request uri https://registry.npmjs.org/browserify
304 verbose request no auth needed
305 info attempt registry request try #1 at 7:05:25 PM
306 verbose etag "E5QLIPZLL2OANON1KH5RX8KXA"
307 http request GET https://registry.npmjs.org/browserify
308 verbose stack TypeError: Cannot read property 'trim' of undefined
308 verbose stack     at C:\Program Files (x86)\nodejs\node_modules\npm\lib\cache\add-remote-git.js:98:31
308 verbose stack     at C:\Program Files (x86)\nodejs\node_modules\npm\lib\utils\git.js:50:14
308 verbose stack     at F (C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\which\which.js:40:25)
308 verbose stack     at E (C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\which\which.js:43:29)
308 verbose stack     at C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\which\which.js:54:16
308 verbose stack     at FSReqWrap.oncomplete (fs.js:95:15)
```

And a screen capture of the actual error message:
![image](https://cloud.githubusercontent.com/assets/4655914/7194925/5b898068-e46c-11e4-8da0-3bca5d00c222.png)

**Comment by KenanY** [comment]
> relevant lines

The _entire_ log is relevant. Could you please upload the entire `npm-debug.log` as a [gist](https://gist.github.com/)?

**Comment by Mr-Shifty** [comment]
Yeah, sorry about that; didn't know about gist. [Here's the log.](https://gist.github.com/Mr-Shifty/41d539f72b7ce3d46231)

And [this](https://github.com/sequitur/raconteur/blob/master/package.json) is the package.json I'm trying to install.

**Comment by dtarandek** [comment]
I just had the same problem on a freshly installed server. Solved the issue by downgrading node.js from 0.12.2 to 0.12.1

```
root@g:~# npm install -g pm2
npm ERR! Linux 3.2.0-4-amd64
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "-g" "pm2"
npm ERR! node v0.12.2
npm ERR! npm  v2.7.4

npm ERR! Cannot read property 'trim' of undefined
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
\
npm ERR! Please include the following file with any support request:
npm ERR!     /root/npm-debug.log
```

```
root@g:~# n 0.12.1

     install : node-v0.12.1
       mkdir : /usr/local/n/versions/node/0.12.1
       fetch : https://nodejs.org/dist/v0.12.1/node-v0.12.1-linux-x64.tar.gz
   installed : v0.12.1
```

```
root@g:~# npm install -g pm2
npm WARN `git config --get remote.origin.url` returned wrong result (http://ikt.pm2.io/ikt.git) undefined
npm WARN `git config --get remote.origin.url` returned wrong result (http://ikt.pm2.io/ikt.git) undefined
npm ERR! git clone http://ikt.pm2.io/ikt.git undefined
npm ERR! git clone http://ikt.pm2.io/ikt.git undefined
npm WARN optional dep failed, continuing ikt@git+http://ikt.pm2.io/ikt.git#master
npm WARN optional dep failed, continuing fsevents@0.3.5
/usr/local/bin/pm2 -> /usr/local/lib/node_modules/pm2/bin/pm2
pm2@0.12.12 /usr/local/lib/node_modules/pm2
├── json-stringify-safe@5.0.0
├── cli-table@0.3.1
├── eventemitter2@0.4.14
├── commander@2.6.0
├── pidusage@0.1.1
├── cron@1.0.5
├── async@0.9.0
├── debug@2.1.0 (ms@0.6.2)
├── isbinaryfile@2.0.3
├── vizion@0.2.6
├── pm2-multimeter@0.1.2 (charm@0.1.2)
├── colors@1.0.3
├── chalk@0.5.1 (ansi-styles@1.1.0, escape-string-regexp@1.0.3, supports-color@0.2.0, has-ansi@0.1.0, strip-ansi@0.3.0)
├── pm2-deploy@0.1.2 (tv4@1.0.18)
├── semver@4.3.3
├── shelljs@0.3.0
├── pm2-axon-rpc@0.3.6 (commander@1.0.5)
├── pmx@0.3.9
├── nssocket@0.5.3 (lazy@1.0.11)
├── pm2-axon@2.0.8 (amp-message@0.1.2, escape-regexp@0.0.1, amp@0.3.1, configurable@0.0.1, debug@2.0.0)
├── coffee-script@1.8.0 (mkdirp@0.3.5)
├── pm2-logs@0.1.1 (chalk@0.4.0, pm2-interface@1.1.0, blessed@0.0.36)
├── moment@2.9.0
└── chokidar@1.0.1 (arrify@1.0.0, is-glob@1.1.3, glob-parent@1.2.0, async-each@0.1.6, is-binary-path@1.0.0, readdirp@1.3.0, anymatch@1.3.0)
```

Upgraded back to 0.12.2 and evertyhing seems to work fine.

**Comment by matthiaswatkins** [comment]
That worked for me. Thanks!

**Comment by Xenoveritas** [comment]
It looks like this was fixed in cd67a0db07891d20871822696c26692c8a84866a and is masking the real error (most likely, you don't have git installed).

In any case, it looks like if it's failing at that specific line with npm 2.7.4 (the version I have from node 0.12.2), the real problem is that npm can't launch git for some reason.

**Comment by liamprescott** [comment]
Hi,

Had a similar issue having updated nodejs and then trying to install the latest version of node-sass (on windows). This was one of a number of error states / messages i picked through before getting it all to work.

Not sure if this will help but basically the issue was caused by two things:
- Missing GIT reference on the PATH
  Being on Windows i'd used the windows installer for Git - *Great except it doesn't add to the windows PATH and 'some' packages seem to need it to be there (some don't...??)*
- Permissions / installing modules with Administrator permissions - *Despite running command prompt with administrator permissions it appears that i didn't have full permission as Windows wasn't releasing the lock on some files*

Basically follow link 1. then restart your machine and try installing.

Here are some useful links:
1. http://stackoverflow.com/questions/26620312/installing-git-in-path-with-github-client-for-windows
2. https://github.com/npm/npm/issues/7528

Hope this helps

**Comment by yeahoffline** [comment]
same here:

$ npm install pm2
npm ERR! Linux 3.13.0-44-generic
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "pm2"
npm ERR! node v0.12.2
npm ERR! npm  v2.7.4

npm ERR! Cannot read property 'trim' of undefined
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
|
npm ERR! Please include the following file with any support request:
npm ERR!     /home/ubuntu/adfusion/npm-debug.log

**Comment by vijayrajasekaran** [comment]
I tried updating pm2 with "npm install pm2@latest -g" (Cent OS 6) but received the following error. Now, i am no longer able to use "pm2 list" (it says command not found).

npm ERR! Linux 2.6.32-042stab094.8
npm ERR! argv "node" "/home/username/local/bin/npm" "install" "pm2" "-g"
npm ERR! node v0.12.1
npm ERR! npm  v2.7.3

npm ERR! Cannot read property 'trim' of undefined
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

**Comment by vijayrajasekaran** [comment]
Updating to Latest npm (2.11.1) solved this issue. 
Thanks.

**Comment by 64BitChris** [comment]
This was just happening for me on ElasticBeanstalk with Node v0.12.2 and npm v2.74.  It's a super vague error and, as @Xenoveritas  mentions it's because git isn't installed on the EB box.  The fix I did was to create an `.ebextensions` directory in my deploy bundle and put a file called `01packages.config` in it with the following contents:

```yml
packages:
  yum:
    git: []
```

I bundled a new zip file and deployed and the problem went away.

If you're using some other server, you can probably just rum `sudo yum install git` or whatever your distro's equivalent is.
